site/_posts/2021-02-14-securing-golang-microservice.md

```diff
@@ -0,0 +1,515 @@
+
+      token, err := jwt.Parse(r.Header["Token"][0], func(token *jwt.Token) (interface{}, error) {
+        if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+          return nil, fmt.Errorf(("Invalid Signing Method"))
+        }
+        aud := "billing.jwtgo.io"
+        checkAudience := token.Claims.(jwt.MapClaims).VerifyAudience(aud, false)
+        if !checkAudience {
+          return nil, fmt.Errorf(("invalid aud"))
+        }
+        // verify iss claim
+        iss := "jwtgo.io"
+        checkIss := token.Claims.(jwt.MapClaims).VerifyIssuer(iss, false)
+        if !checkIss {
+          return nil, fmt.Errorf(("invalid iss"))
+        }
+
+        return MySigningKey, nil
+      })
+      if err != nil {
+        fmt.Fprintf(w, err.Error())
+      }
+
+      if token.Valid {
+        endpoint(w, r)
+      }
+
+    } else {
+      fmt.Fprintf(w, "No Authorization Token provided")
+    }
+  })
+}
+func handleRequests() {
+  http.Handle("/", isAuthorized(homePage))
+  log.Fatal(http.ListenAndServe(":9001", nil))
+}
+func main() {
+  fmt.Println("server")
+  handleRequests()
+}
+```
+
+### Start the microservice
+
+We need to start the middleware and microservice server by executing the following command:
+
+```shell
+go run main.go
+```
+
+Now we can test the endpoint with curl. First, try without any JWT:
+
+```shell
+curl http://localhost:9001
+```
+
+You'll receive an error message:
+
+```
+No Authorization Token provided
+```
+
+Next, generate a token:
+
+```
+curl http://localhost:8080 
+```
+
+You'll see something like this:
+
+```
+eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJiaWxsaW5nLmp3dGdvLmlvIiwiYXV0aG9yaXplZCI6dHJ1ZSwiY2xpZW50IjoiS3Jpc3NhbmF3YXQiLCJleHAiOjE2MTM1MDk1MDcsImlzcyI6Imp3dGdvLmlvIn0.t7qdqrpLk3nBOZFLBL_UOdciZ_rWei0rJg3tgyJ7cTw
+```
+
+Now we can construct the curl command as shown below:
+
+```shell
+curl http://localhost:9001 --header 'Token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJiaWxsaW5nLmp3dGdvLmlvIiwiYXV0aG9yaXplZCI6dHJ1ZSwiY2xpZW50IjoiS3Jpc3NhbmF3YXQiLCJleHAiOjE2MTM1MDk1MDcsImlzcyI6Imp3dGdvLmlvIn0.t7qdqrpLk3nBOZFLBL_UOdciZ_rWei0rJg3tgyJ7cTw'
+```
+
+If you execute this request this within a minute of when you created the token, you'll be rewarded with:
+
+```
+Super Secret Information
+```
+
+If you take too long, you'll see a token expiration warning. And no super secret information.
+
+## Conclusion
+
+This post demonstrated the generation of a JWT in golang. It used the same token to access a golang microservice running in another server. The token generation used the `go-jwt` module with the HS256 signing method. The generated token was returned to the client after requesting an endpoint. 
+
+Then, for the microservices server, middleware logic validated the token. After successful validation the request was sent to the microservices endpoint. 
+
+Using a token in this manner to secure microservices is widely used as it is a secure and easy mechanism. 
+
+## Go further
```

Comment:
Don't know if this makes sense, but we could point to the "Securing A Golang App with OAuth" blog post?

Reply:
Works for me, I think it'd be interesting.